packages/world-modules/src/modules/delegation/DelegationSystem.sol

```diff
@@ -0,0 +1,47 @@
+// SPDX-License-Identifier: MIT
+pragma solidity >=0.8.24;
+
+import { ResourceId } from "@latticexyz/store/src/ResourceId.sol";
+import { System } from "@latticexyz/world/src/System.sol";
+import { createDelegation } from "@latticexyz/world/src/modules/init/implementations/createDelegation.sol";
+
+import { UserDelegationNonces } from "./tables/UserDelegationNonces.sol";
+import { getSignedMessageHash } from "./getSignedMessageHash.sol";
+import { ECDSA } from "./ECDSA.sol";
+
+contract DelegationSystem is System {
+  /**
+   * @dev Mismatched signature.
+   */
+  error InvalidSigner(address delegator, address delegatee);
+
+  /**
+   * @notice Registers a delegation for `delegator` with a signature
+   * @dev Creates a new delegation from the caller to the specified delegatee
+   * @param delegatee The address of the delegatee
+   * @param delegationControlId The ID controlling the delegation
+   * @param initCallData The initialization data for the delegation
+   * @param delegator The address of the delegator
+   * @param signature The EIP712 signature
+   */
+  function registerDelegationWithSignature(
+    address delegatee,
+    ResourceId delegationControlId,
+    bytes memory initCallData,
+    address delegator,
+    bytes memory signature
+  ) public {
+    uint256 nonce = UserDelegationNonces.get(delegator);
+    bytes32 hash = getSignedMessageHash(delegatee, delegationControlId, initCallData, nonce, _world());
```

Comment:
should the `delegator` be part of the signed message?

Reply:
Good question. My intuition is no because the `delegator` is always the signer. 

If someone submits a signature that was signed by an different account to the provided`delegator`, `ecrecover` will fail. We don't ALSO need to put it in the signature.

Reply:
Do we need `delegator` as an additional argument in the function signature then? Couldn't we just always use the recovered `signer`, and just revert if it's `address(0)`?

Btw I think right now it's possible to submit `address(0)` as the `delegator` param and successfully register a signature for `address(0)` because `ECDSA.tryRecover` would return `address(0)` if the signature is invalid.

Reply:
I'd say it's good practice to include `delegator` param in this case to make sure that the function is behaving as intended i.e. signature resolves to the expected delegator, or otherwise reverts. Besides that can't think of other good reasons right now though so don't think it's strictly necessary

Reply:
We should really use `recover`, the reason I didn't before is because I wanted to emit a `World_` error instead of the `ECDSA` one

Reply:
> signature resolves to the expected delegator

My understanding is it's impossible to create a valid signature that resolves to a different delegator? If it's possible to accidentally create a signature that resolves to some signer who didn't actually sign it, then i feel like we should include the `delegator` in the signed message to prevent that from accidentally happening. (_If_ this was possible, it would be really easy to locally recover the "signer" and pass it as `delegator` param to register a delegation on behalf of someone who didn't actually sign this)

Reply:
Yeah, we can assume it's not possible. The comment about "expected delegator" is related more to the consumer side to make sure that their message signing implementation is correct, and the signatures they produce resolve to the delegator that is expected.

Reply:
We synced offline, putting it here for context:
- an "invalid signature" would "recover" _some_ address, probably one without a known private key
- it would be possible to pass this address as the delegator and thereby register a delegation for an address that never signed the delegation
- to avoid that we can make the delegator part of the message that is being signed, so it's not possible to create an invalid signature, recover a signer, and use the signer as delegator.